include/deal.II/base/thread_local_storage.h

```diff
@@ -158,29 +153,25 @@ namespace Threads
     void
     clear();
 
+  private:
     /**
-     * Return a reference to the internal Threading Building Blocks
-     * implementation. This function is really only useful if deal.II has been
-     * configured with multithreading and has no useful purpose otherwise.
+     * The data element we store.
      */
-#  ifdef DEAL_II_WITH_THREADS
-    tbb::enumerable_thread_specific<T> &
-#  else
-    T &
-#  endif
-    get_implementation();
+    std::map<std::thread::id, T> data;
 
-  private:
-#  ifdef DEAL_II_WITH_THREADS
     /**
-     * The data element we store. If we support threads, then this object will
-     * be of a type that provides a separate object for each thread.
-     * Otherwise, it is simply a single object of type T.
+     * A mutex to guard insertion into the data object.
      */
-    tbb::enumerable_thread_specific<T> data;
+#  ifdef DEAL_II_HAVE_CXX17
+    std::shared_mutex insertion_mutex;
 #  else
-    T data;
+    std::shared_timed_mutex insertion_mutex;
 #  endif
+
+    /**
+     * An exemplar for creating a new (thread specific) copy.
+     */
+    std::shared_ptr<T> exemplar;
```

Comment:
Could we make this `std::shared_ptr<const T>` to make reasoning about why a shared pointer is the right approach here easier? Intuitively, I wanted to make this a `unique_ptr`, but then realized that `shared_ptr` is entirely sufficient -- but really only because no thread will ever change the exemplar.